### DISCUSSION TITLE
Need help about the fluentd memory leak issue

### DISCUSSION THREAD

**Discussion by sunnystone85** [fluent/fluentd#4561] [General]
I am a new about this.

Upgraded the fluentd from td-agent v4 to fluent-package v5, after upgrade to the new version around 2 weeks, the memory reached > 90%.
We need to restart the services.

There are two ruby plugin, in the match configuartion.

I have no clue about the issue, how I can deal with this issue.

![image](https://github.com/user-attachments/assets/f1ad05ce-8441-4bca-b384-a48e7c71ded7)

**Comment by daipom** [comment]
@sunnystone85 Sorry for the trouble.
Could you share the config?

**Reply by sunnystone85** [reply]
Thank you for you reply.

```
system {
  root_dir "/usr/local/xxxx/tools/fluentd/backup_root"
}

source {
  type     :tail
  tag      "xxxxx.log"
  path     "/var/log/httpd/xxxxxx/access.log"
  pos_file "pos/xxxx.pos"
  format   '/^"(?<useragent>[^"]+)" "(?<server_id>[0-9]+)" "t=(?<request_time_u>[0-9]+)" "(?<histid>[^"]+)" "(?<uid>[^"]+)" "(?<tracking_status>[0-9]+)" "(?<request_line>[a-zA-Z]+\s\/(c|s|st)\/[^"]+)" (?<status>[0-9]+).+?"(?<referer>[^"]*)" [0-9.]+ [0-9]+ "[^"]+" (?<remote_ip>[0-9.]+) "(?<rz_cookie>[^"]+)" "(?<rp_cookie>[^"]+)" "(?<ra_uid_cookie>[^"]+)"/'
  rotate_wait "30s"
}

match ('{xxxxx}.**') {
  tag "xx"
  type :exec_filter
  in_format :json
  out_format :msgpack
  command "ruby ./scripts/xxxxx.rb"
  flush_interval "1s"
  buffer_type :file
  buffer_path "buffer/xxxx.*.buffer"
  buffer_chunk_limit "256m"  # default: 256m
  buffer_queue_limit "128"   # default: 128
  flush_at_shutdown :false   # default: false
  disable_retry_limit :true  # default: 17
  retry_wait "1s"            # default: 1s
}

match ('{xxxx}.**') {
  type :file
  path "unmatched/xxx.*"
  compress :gz
}

source {
  type :monitor_agent
  bind "0.0.0.0"
  port 24220
}

match ('**') {
  type :file
  path "unmatched/xxxxxx_access_log.*"
  compress :gz
}
```

**Reply by daipom** [reply]
Thanks!
I would like to know which plugin is causing the problem.

Is it possible to change the setting and test?
If it is, could you try removing `exec_filter` and see how it goes?

**Reply by sunnystone85** [reply]
So you mean, it should be the plugin script cause the problem right?
If it is caused by the script
   command "ruby ./scripts/xxxxx.rb"

If so how we can fix the issue? the new ruby version not compatible with the old code?
Could you provide some suggestion?

**Reply by daipom** [reply]
> So you mean, it should be the plugin script cause the problem right?

I'm not sure, but it would be good to narrow down the possibilities.
It could also be the problem of `out_exec_filter` plugin itself.
Or, it could also be the problem of another plugin, such as `in_tail`, `out_file`, or `buffer`.

If we can reproduce the problem simply, we can investigate the cause.

> If so how we can fix the issue? the new ruby version not compatible with the old code?

If we conclude that the script is the cause, maybe someone, including me, can figure out something if you could share the script to the extent possible.

**Reply by sunnystone85** [reply]
Thank you, will try to reproduce the issue, then let you know.
BTW, there is a while section in the plugin script, the script will keep running and causing memory leak?

```rb
while str = STDIN.gets
  xxxxxxxxxxxxxxxxxx
  $log.info(r)
  next if r.empty?
  print MessagePack.pack(r)
  STDOUT.flush
end
```

**Reply by daipom** [reply]
The script looks fine.
It should keep running with Fluentd.
If there is more than one script running, something is wrong.

**Reply by daipom** [reply]
Are the logs outputted correctly?
If not, some data may have accumulated without being outputted.
